src/search/index_info.h

```diff
@@ -52,21 +53,28 @@ struct FieldInfo {
 
 struct IndexInfo {
   using FieldMap = std::map<std::string, FieldInfo>;
+  using MutexMap = std::map<std::string, std::mutex>;
```

Comment:
Maybe we can put mutexes into `IndexUpdater`s instead of `IndexInfo`s?

Reply:
Sure, this makes more sense <3

Reply:
While trying to move the exact same piece of code into `IndexUpdater`s, I encountered the following errors:
```
‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const std::__cxx11::basic_string<char>; _T2 = std::mutex]’ is implicitly deleted because the default definition would be ill-formed:
  314 |       constexpr pair(const pair&) = default;    ///< Copy constructor
```
I suspect that this is caused by `IndexUpdater`s are usually created on stacks, which causes `std::map` doing some underlying copies and deletions, while `IndexInfo`s are allocated on heap so doesn't have the issue. 
I tried to wrap `std::mutex` into a `unique_ptr` but the compiler will report the same error. 

I'm not quite sure what I can do at this point, may leave it as it is, and try to benchmark first. But feel free to let me know if you think there is anything I missed, or any workaround you can think of. @PragmaTwice  <3

Reply:
I think we can try `map<.., unique_ptr<mutex>>` to avoid such problem.

Reply:
Sorry for the delayed response.

I checked the related code, and I think there's two methods to solve it:
1. keep mutex maps inside `GlobalIndexer`, and add a raw pointer into `IndexUpdater` to access them;
2. follow the current way and modify `void Add(IndexUpdater updater)` to something like `Add(std::unqiue_ptr<IndexUpdater>)`.

Reply:
All good all good, thanks for sharing your solution! 
Might be a bit busy these two weeks, will be back to work on it during mid October.